reactos/drivers/storage/storahci/storahci.c

```diff
@@ -0,0 +1,745 @@
+/*
+ * PROJECT:        ReactOS Kernel
+ * LICENSE:        GNU GPLv2 only as published by the Free Software Foundation
+ * PURPOSE:        To Implement AHCI Miniport driver targeting storport NT 5.2
+ * PROGRAMMERS:    Aman Priyadarshi (aman.eureka@gmail.com)
+ */
+
+#include "storahci.h"
+
+/**
+ * @name AhciPortInitialize
+ * @implemented
+ *
+ * Initialize port by setting up PxCLB & PxFB Registers
+ *
+ * @param portExtension
+ *
+ * @return
+ * Return true if intialization was successful
+ */
+BOOLEAN AhciPortInitialize(
+  __in      PAHCI_PORT_EXTENSION                portExtension
+)
+{
+    ULONG mappedLength;
+    PAHCI_MEMORY_REGISTERS abar;
+    PAHCI_ADAPTER_EXTENSION adapterExtension;
+    STOR_PHYSICAL_ADDRESS commandListPhysical, receivedFISPhysical;
+
+    StorPortDebugPrint(0, "AhciPortInitialize()\n");
+
+    adapterExtension = portExtension->AdapterExtension;
+    abar = adapterExtension->ABAR_Address;
+
+    portExtension->Port = &abar->PortList[portExtension->PortNumber];

```

Comment:
Would be nice to see some NT_ASSERTs, such as making sure PortNumber is not too big, that abar isn't NULL, etc..